### PR TITLE
DM requests for invoices

### PR DESCRIPTION
**Pull request by benthecarman**
Think this is all the functionality needed for @futurepaul to start

### CODE REVIEW

File: mutiny-core/src/lib.rs [hunk 1]
```diff
@@ -629,7 +629,7 @@ impl<S: MutinyStorage> MutinyWalletBuilder<S> {
 pub struct MutinyWallet<S: MutinyStorage> {
     xprivkey: ExtendedPrivKey,
     config: MutinyWalletConfig,
-    pub(crate) storage: S,
+    pub storage: S,
```
Comment (TonyGiorgio) [inline mutiny-core/src/lib.rs:632]: Pass storage to things that need storage.

**Review by TonyGiorgio — CHANGES_REQUESTED** [review]

File: mutiny-core/src/storage.rs [hunk 2]
```diff
@@ -268,6 +273,28 @@ pub trait MutinyStorage: Clone + Sized + Send + Sync + 'static {
         self.get_data(MNEMONIC_KEY)
     }
 
+    /// Insert a nsec into the storage
+    async fn insert_nsec(&self, nsec: SecretKey) -> Result<(), MutinyError> {
```
Comment (TonyGiorgio) [inline mutiny-core/src/storage.rs:277]: Hard NACK. Nsec should not be pasted into a browser site.

**Review by TonyGiorgio — CHANGES_REQUESTED** [review]

File: mutiny-core/src/storage.rs (hunk 2 above)
Comment (benthecarman) [reply mutiny-core/src/storage.rs:277]: Should discuss IRL but paul and I's thinking was we already doing worse with a bitcoin seed

Comment (TonyGiorgio) [reply mutiny-core/src/storage.rs:277]: Huge difference in a mutiny seed being used for mutiny things and importing another seed/private key from somewhere else for other purposes. 

Find another way, there's others that exist. Maybe ask Pablo...

Comment (TonyGiorgio) [reply mutiny-core/src/storage.rs:277]: There's a few things I think we can do here but I would absolutely not allow nsec pasting here. 

1. Nostr plugins for web browsers that already have them
2. Create an nsec from the derivation path and use that for users that don't already have nostr
  2a. Some delegated nostr stuff so you can use the mutiny derivation seed as if it were the user's main key (don't think this really exists yet)
3. nsec bunker for everyone else

**Comment by benthecarman** [comment]
Changed to not import but be instead be part of the building phase. We can then instead store it somewhere more secure

**Comment by TonyGiorgio** [comment]
Needs clippy fixes

File: mutiny-core/src/nostr/mod.rs [hunk 3]
```diff
@@ -91,14 +106,17 @@ impl<S: MutinyStorage> NostrManager<S> {
             .map(|x| x.profile.relay.clone())
             .collect();
 
+        // add relay to pull DMs from
+        relays.push("wss://relay.primal.net".to_string());
```
Comment (TonyGiorgio) [inline mutiny-core/src/nostr/mod.rs:110]: add damus

File: mutiny-core/src/storage.rs [hunk 4]
```diff
@@ -27,6 +27,7 @@ use uuid::Uuid;
 
 pub const KEYCHAIN_STORE_KEY: &str = "bdk_keychain";
 pub const MNEMONIC_KEY: &str = "mnemonic";
+pub const NSEC_KEY: &str = "nsec";
```
Comment (TonyGiorgio) [inline mutiny-core/src/storage.rs:30]: delete this

**Review by TonyGiorgio — CHANGES_REQUESTED** [review]

File: mutiny-core/src/lib.rs [hunk 5]
```diff
@@ -1420,6 +1462,64 @@ impl<S: MutinyStorage> MutinyWallet<S> {
...
+    /// Get dm conversation between us and given npub
+    /// Returns a vector of messages sorted by newest first
+    pub async fn get_dm_conversation(
+        &self,
+        primal_url: Option<&str>,
+        npub: XOnlyPublicKey,
+        limit: u64,
+        until: Option<u64>,
+    ) -> Result<Vec<DirectMessage>, MutinyError> {
+        let url = primal_url.unwrap_or("https://primal-cache.mutinywallet.com/api");
```
[the remark this answers is not in the document]
Comment (benthecarman) [reply mutiny-core/src/lib.rs:1474]: Currently we have it an an ENV in the front end and pass it for the individual requests, probably could refactor to have this as part of init, maybe good for a follow up so we can do it with the sync nostr contacts stuff too

File: mutiny-core/src/nostr/mod.rs [hunk 6]
```diff
@@ -1011,14 +1183,33 @@ impl<S: MutinyStorage> NostrManager<S> {
...
         stop: Arc<AtomicBool>,
     ) -> Result<Self, MutinyError> {
         let context = Secp256k1::new();
 
-        // generate the default primary key
-        let primary_key = Self::derive_nostr_key(&context, xprivkey, 0, None, None)?;
+        // use provided nsec, otherwise generate it from seed
+        let (primary_key, public_key) = match key_source {
+            NostrKeySource::Derived => {
+                let keys = Self::derive_nostr_key(&context, xprivkey, 0, None, None)?;
```
[the remark this answers is not in the document]
Comment (benthecarman) [reply mutiny-core/src/nostr/mod.rs:1196]: Yeah we're using `0` to comply with nip6, will make it an const

https://github.com/nostr-protocol/nips/blob/master/06.md

File: mutiny-core/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (TonyGiorgio) [reply mutiny-core/src/lib.rs:1474]: We're already changing the mutiny wallet init parameters, might as well add it in this PR

**Review by TonyGiorgio — APPROVED** [review]

**Comment by benthecarman** [comment]
Was missing send functionality and an efficient way to get future messages. Added that. Cleaned up some of our nostr client handling as well so we don't need to make so many new connections

**Comment by TonyGiorgio** [comment]
Tests are panicing

**Review by TonyGiorgio — APPROVED** [review]

**Pull request merged**